src/store.cc

```diff
@@ -1537,7 +1537,12 @@ StoreEntry::updateOnNotModified(const StoreEntry &e304)
     if (!timestampsSet() && !updatedReply)
         return false;
 
-    // XXX: Update old->mem_obj->vary_headers?
+    assert(e304.mem_obj->request);
+    const auto updatedVaryMark = updatedReply->makeVaryMark(*(e304.mem_obj->request));
+    if (mem_obj->vary_headers != updatedVaryMark) {
+        debugs(20, 2, "unexpected vary headers change in the not-modified response detected!");
+        mem_obj->vary_headers = updatedVaryMark;
```

Comment:
Updating MemObject::vary_headers is not enough for everything to continue to work OK, right? The updated object will be stored, but no future request will find it because the vary switch entry will still have the old list of Vary-controlled request header names, right? Perhaps it is best _not_ to update `vary_headers` in this case?

We can refuse to fix other problems if they are out of scope, but we should at least disclose them. And we should fix what is easy to fix (if anything). Squid forgets the very switch entry after finding it so we cannot even invalidate it here, right? Invaliding it may not be a good idea in itself, but that is a separate question.

Reply:
> Perhaps it is best not to update vary_headers in this case?

I think we should update it anyway: it should match the updated vary headers, stored in the reply. IMO, the major problem here is different and exists regardless of updating/not_updating this (secondary) field. After update, the entry de facto acquires a new key (since vary changed), while still being attached to the old key (via store_table). So, further requests with new vary-related header will miss, while requests with old vary-related header will hit (both are wrong).